File: app/src/main/java/ru/melodin/fast/api/model/VKAttachments.java
package ru.melodin.fast.api.model;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.Serializable;
import java.util.ArrayList;

import ru.melodin.fast.R;
import ru.melodin.fast.util.ArrayUtil;

public class VKAttachments extends VKModel implements Serializable {

    public static final String TYPE_PHOTO = "photo";
    public static final String TYPE_VIDEO = "video";
    public static final String TYPE_AUDIO = "audio";
    public static final String TYPE_DOC = "doc";
    public static final String TYPE_WALL = "wall";
    public static final String TYPE_POSTED_PHOTO = "posted_photo";
    public static final String TYPE_LINK = "link";
    public static final String TYPE_NOTE = "note";
    public static final String TYPE_APP = "app";
    public static final String TYPE_POLL = "poll";
    public static final String TYPE_WIKI_PAGE = "page";
    public static final String TYPE_ALBUM = "album";
    public static final String TYPE_STICKER = "sticker";
    public static final String TYPE_GIFT = "gift";
    public static final String TYPE_AUDIO_MESSAGE = "audio_message";
    public static final String TYPE_GRAFFITI = "graffiti";

    public static ArrayList<VKModel> parse(JSONArray array) {
        ArrayList<VKModel> attachments = new ArrayList<>(array.length());

        for (int i = 0; i < array.length(); i++) {
            JSONObject attachment = array.optJSONObject(i);
            if (attachment.has("attachment")) {
                attachment = attachment.optJSONObject("attachment");
            }

            String type = attachment.optString("type");
            JSONObject object = attachment.optJSONObject(type);

            switch (type) {
                case TYPE_PHOTO:
                    attachments.add(new VKPhoto(object));
                    break;
                case TYPE_AUDIO:
                    attachments.add(new VKAudio(object));
                    break;
                case TYPE_VIDEO:
                    attachments.add(new VKVideo(object));
                    break;
                case TYPE_DOC:
                    attachments.add(new VKDoc(object));
                    break;
                case TYPE_STICKER:
                    attachments.add(new VKSticker(object));
                    break;
                case TYPE_LINK:
                    attachments.add(new VKLink(object));
                    break;
                case TYPE_GIFT:
                    attachments.add(new VKGift(object));
                    break;
                case TYPE_AUDIO_MESSAGE:
                    attachments.add(new VKVoice(object));
                    break;
                case TYPE_GRAFFITI:
                    attachments.add(new VKGraffiti(object));
                    break;
                case TYPE_WALL:
                    attachments.add(new VKWall(object));
                    break;
            }
        }

        return attachments;
    }

    public static String getAttachmentString(ArrayList<VKModel> attachments) {
        if (ArrayUtil.isEmpty(attachments)) return "";
        StringBuilder b = new StringBuilder();

        if (attachments.size() > 1) {
            return attachments.size() + " " + getString(R.string.attachments_lot).toLowerCase();
        }

        for (VKModel attach : attachments) {
            if (attach instanceof VKAudio) {
                b.append(getString(R.string.audio));
            } else if (attach instanceof VKPhoto) {
                b.append(getString(R.string.photo));
            } else if (attach instanceof VKSticker) {
                b.append(getString(R.string.sticker));
            } else if (attach instanceof VKDoc) {
                b.append(getString(R.string.doc));
            } else if (attach instanceof VKLink) {
                b.append(getString(R.string.link));
            } else if (attach instanceof VKVideo) {
                b.append(getString(R.string.video));
            } else if (attach instanceof VKVoice) {
                b.append(getString(R.string.voice_message));
            } else if (attach instanceof VKGraffiti) {
                b.append(getString(R.string.graffiti));
            } else if (attach instanceof VKGift) {
                b.append(getString(R.string.gift));
            } else if (attach instanceof VKWall) {
                b.append(getString(R.string.wall_post));
            } else {
                b.append(getString(R.string.attachment));
            }
        }

        return b.toString();
    }

    private static boolean isOneType(Class type, ArrayList<VKModel> attachments) {
        if (ArrayUtil.isEmpty(attachments) || type == null) return false;

        for (VKModel a : attachments) {
            if (!a.getClass().equals(type)) {
                return false;
            }
        }

        return true;
    }

    public static ArrayList<VKModel> parseFromLongPoll(JSONObject o) {
        ArrayList<VKModel> attachments = new ArrayList<>();

        for (int i = 0; i < o.length() / 2; i++) {
            String a_type = "attach" + i + "_type";
            String attach = "attach" + i;

            if (o.has(a_type)) {
                String type = o.optString(a_type);

                String attachment = o.optString(attach);
                String[] s = attachment.split("_");

                if (ArrayUtil.isEmpty(s) || s.length == 1) return null;

                int peerId = Integer.parseInt(s[0]);
                int attId = Integer.parseInt(s[1]);

                switch (type) {
                    case TYPE_PHOTO:
                        attachments.add(new VKPhoto(peerId, attId));
                        break;
                    case TYPE_AUDIO:
                        attachments.add(new VKAudio(peerId, attId));
                        break;
                    case TYPE_VIDEO:
                        attachments.add(new VKVideo(peerId, attId));
                        break;
                    case TYPE_DOC:
                        attachments.add(new VKDoc(peerId, attId));
                        break;
                    case TYPE_STICKER:
                        attachments.add(new VKSticker(peerId, attId));
                        break;
                    case TYPE_LINK:
                        attachments.add(new VKLink());
                        break;
                    case TYPE_GIFT:
                        attachments.add(new VKGift(peerId, attId));
                        break;
                }
            }
        }

        return attachments;
    }
}

File: app/src/main/java/ru/melodin/fast/api/model/VKAttachments.kt
package ru.melodin.fast.api.model

import android.annotation.SuppressLint
import androidx.annotation.StringRes
import org.json.JSONArray
import ru.melodin.fast.R
import ru.melodin.fast.api.model.attachment.*
import ru.melodin.fast.common.AppGlobal
import ru.melodin.fast.util.ArrayUtil
import java.io.Serializable
import java.util.*

class VKAttachments : VKModel(), Serializable {
    companion object {

        private const val TYPE_PHOTO = "photo"
        private const val TYPE_VIDEO = "video"
        private const val TYPE_AUDIO = "audio"
        private const val TYPE_DOC = "doc"
        private const val TYPE_WALL = "wall"
        private const val TYPE_POSTED_PHOTO = "posted_photo"
        private const val TYPE_LINK = "link"
        private const val TYPE_NOTE = "note"
        private const val TYPE_APP = "app"
        private const val TYPE_POLL = "poll"
        private const val TYPE_WIKI_PAGE = "page"
        private const val TYPE_ALBUM = "album"
        private const val TYPE_STICKER = "sticker"
        private const val TYPE_STORY = "story"
        private const val TYPE_GIFT = "gift"
        private const val TYPE_AUDIO_MESSAGE = "audio_message"
        private const val TYPE_GRAFFITI = "graffiti"

        fun parse(array: JSONArray): ArrayList<VKModel> {
            val attachments = ArrayList<VKModel>(array.length())

            for (i in 0 until array.length()) {
                var attachment = array.optJSONObject(i)!!

                if (attachment.has("attachment")) {
                    attachment = attachment.optJSONObject("attachment")!!
                }

                val type = attachment.optString("type")
                val source = attachment.optJSONObject(type) ?: return attachments

                when (type) {
                    TYPE_STORY -> attachments.add(VKStory(source))
                    TYPE_PHOTO -> attachments.add(VKPhoto(source))
                    TYPE_AUDIO -> attachments.add(VKAudio(source))
                    TYPE_VIDEO -> attachments.add(VKVideo(source))
                    TYPE_DOC -> attachments.add(VKDoc(source))
                    TYPE_STICKER -> attachments.add(VKSticker(source))
                    TYPE_LINK -> attachments.add(VKLink(source))
                    TYPE_GIFT -> attachments.add(VKGift(source))
                    TYPE_AUDIO_MESSAGE -> attachments.add(VKVoice(source))
                    TYPE_GRAFFITI -> attachments.add(VKGraffiti(source))
                    TYPE_WALL -> attachments.add(VKWall(source))
                }
            }

            return attachments
        }

        @SuppressLint("DefaultLocale")
        fun getAttachmentString(attachments: ArrayList<VKModel>): String {
            if (ArrayUtil.isEmpty(attachments)) return ""
            val b = StringBuilder()

            if (attachments.size > 1) {
                return attachments.size.toString() + " " + getString(R.string.attachments_lot).toLowerCase()
            }

            for (attach in attachments) {
                when (attach) {
                    is VKAudio -> b.append(getString(R.string.audio))
                    is VKPhoto -> b.append(getString(R.string.photo))
                    is VKSticker -> b.append(getString(R.string.sticker))
                    is VKDoc -> b.append(getString(R.string.doc))
                    is VKLink -> b.append(getString(R.string.link))
                    is VKVideo -> b.append(getString(R.string.video))
                    is VKVoice -> b.append(getString(R.string.voice_message))
                    is VKGraffiti -> b.append(getString(R.string.graffiti))
                    is VKGift -> b.append(getString(R.string.gift))
                    is VKWall -> b.append(getString(R.string.wall_post))
                    else -> b.append(getString(R.string.attachment))
                }
            }

            return b.toString()
        }

        private fun isOneType(type: Class<*>?, attachments: ArrayList<VKModel>): Boolean {
            if (ArrayUtil.isEmpty(attachments) || type == null) return false

            for (a in attachments) {
                if (a.javaClass != type) {
                    return false
                }
            }

            return true
        }

        private fun getString(@StringRes resId: Int): String {
            return AppGlobal.res.getString(resId)
        }
    }
}
